### ISSUE TITLE
Fatal error after reload when configuration got changed (increasing sample_limit)

### ISSUE THREAD

**Issue by hawran** [prometheus/prometheus#6997]
**What did you do?**
A configuration has been changed:
`sample_limit` from `100000` to `130000`
then the new configuration has been deployed to kubernetes.

**What did you expect to see?**
A configuration reload as usual.

**What did you see instead? Under which circumstances?**
Prometheus got restarted with a bunch of errors within logs.

**Environment**

* System information:
```
 Kernel Version:                  5.3.0-24-generic
 OS Image:                        Ubuntu 18.04.1 LTS
 Operating System:                linux
 Architecture:                    amd64
 Kubelet Version:                 v1.14.8
```

* Prometheus version:
2.16.0

* Logs:
```
fatal error: concurrent map read and map write

goroutine 7926 [running]:
runtime.throw(0x25d4f3b, 0x21)
	/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-go-1.12.16/share/go/src/runtime/panic.go:617 +0x72 fp=0xc0248e16c0 sp=0xc0248e1690 pc=0x42f552
runtime.mapaccess2_fast64(0x217f780, 0xc21aaea3c0, 0x329a094c9e427070, 0x45dea01, 0xc256bc5d00)
	/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-go-1.12.16/share/go/src/runtime/map_fast64.go:61 +0x1c2 fp=0xc0248e16e8 sp=0xc0248e16c0 pc=0x412ed2
github.com/prometheus/prometheus/scrape.(*scrapeCache).forEachStale(0xc21aaf0100, 0xc0248e1a10)
	/build/go/src/github.com/prometheus/prometheus/scrape/scrape.go:766 +0xdf fp=0xc0248e1798 sp=0xc0248e16e8 pc=0x1d7fbff
github.com/prometheus/prometheus/scrape.(*scrapeLoop).append(0xc0f1760240, 0x45dd750, 0x0, 0x0, 0x0, 0x0, 0xbf94458d1f2e81b8, 0x6181c1af2c1ca, 0x45bf9e0, 0x0, ...)
	/build/go/src/github.com/prometheus/prometheus/scrape/scrape.go:1219 +0xce5 fp=0xc0248e1a48 sp=0xc0248e1798 pc=0x1d83825
github.com/prometheus/prometheus/scrape.(*scrapeLoop).endOfRunStaleness(0xc0f1760240, 0xbf9445895f2e5033, 0x618189ce0ba36, 0x45bf9e0, 0xc16d72bf90, 0x37e11d600)
	/build/go/src/github.com/prometheus/prometheus/scrape/scrape.go:1046 +0x36b fp=0xc0248e1c58 sp=0xc0248e1a48 pc=0x1d8260b
github.com/prometheus/prometheus/scrape.(*scrapeLoop).run(0xc0f1760240, 0x37e11d600, 0x2540be400, 0x0)
	/build/go/src/github.com/prometheus/prometheus/scrape/scrape.go:1003 +0x929 fp=0xc0248e1fc0 sp=0xc0248e1c58 pc=0x1d816d9
runtime.goexit()
	/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-go-1.12.16/share/go/src/runtime/asm_amd64.s:1337 +0x1 fp=0xc0248e1fc8 sp=0xc0248e1fc0 pc=0x45e8f1
created by github.com/prometheus/prometheus/scrape.(*scrapePool).sync
	/build/go/src/github.com/prometheus/prometheus/scrape/scrape.go:425 +0x52c

goroutine 1 [chan receive, 28589 minutes]:
github.com/prometheus/prometheus/vendor/github.com/oklog/run.(*Group).Run(0xc0009b7c38, 0xc000212410, 0x8)
	/build/go/src/github.com/prometheus/prometheus/vendor/github.com/oklog/run/group.go:43 +0xf0
main.main()
	/build/go/src/github.com/prometheus/prometheus/cmd/prometheus/main.go:739 +0x6edd

goroutine 109 [IO wait, 28589 minutes]:
internal/poll.runtime_pollWait(0x7f46bd265de8, 0x72, 0xffffffffffffffff)
	/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-go-1.12.16/share/go/src/runtime/netpoll.go:182 +0x56
internal/poll.(*pollDesc).wait(0xc0005ae198, 0x72, 0x8000, 0x8000, 0xffffffffffffffff)
	/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-go-1.12.16/share/go/src/internal/poll/fd_poll_runtime.go:87 +0x9b
internal/poll.(*pollDesc).waitRead(...)
	/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-go-1.12.16/share/go/src/internal/poll/fd_poll_runtime.go:92
internal/poll.(*FD).Read(0xc0005ae180, 0xc0009be000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-go-1.12.16/share/go/src/internal/poll/fd_unix.go:169 +0x19b
net.(*netFD).Read(0xc0005ae180, 0xc0009be000, 0x8000, 0x8000, 0x45b3a0, 0xc000554300, 0x4)
	/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-go-1.12.16/share/go/src/net/fd_unix.go:202 +0x4f
net.(*conn).Read(0xc000010028, 0xc0009be000, 0x8000, 0x8000, 0x0, 0x0, 0x0)
	/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-go-1.12.16/share/go/src/net/net.go:177 +0x69
bufio.(*Reader).Read(0xc00015ea80, 0xc0004921f8, 0x9, 0x9, 0x4591a0, 0xc00073c6c0, 0xc0000705f8)
	/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-go-1.12.16/share/go/src/bufio/bufio.go:223 +0x23e
io.ReadAtLeast(0x2bed900, 0xc00015ea80, 0xc0004921f8, 0x9, 0x9, 0x9, 0x1000000022d6ce0, 0xc0004a00e0, 0xc0007400d0)
	/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-go-1.12.16/share/go/src/io/io.go:310 +0x88
io.ReadFull(...)
	/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-go-1.12.16/share/go/src/io/io.go:329
github.com/prometheus/prometheus/vendor/golang.org/x/net/http2.readFrameHeader(0xc0004921f8, 0x9, 0x9, 0x2bed900, 0xc00015ea80, 0x0, 0xc000000000, 0xc0007031c0, 0xc00066c910)
	/build/go/src/github.com/prometheus/prometheus/vendor/golang.org/x/net/http2/frame.go:237 +0x88
github.com/prometheus/prometheus/vendor/golang.org/x/net/http2.(*Framer).ReadFrame(0xc0004921c0, 0xc000542150, 0x1, 0x0, 0x0)
	/build/go/src/github.com/prometheus/prometheus/vendor/golang.org/x/net/http2/frame.go:492 +0xa1
github.com/prometheus/prometheus/vendor/google.golang.org/grpc/internal/transport.(*http2Client).reader(0xc000432540)
	/build/go/src/github.com/prometheus/prometheus/vendor/google.golang.org/grpc/internal/transport/http2_client.go:1242 +0x168
created by github.com/prometheus/prometheus/vendor/google.golang.org/grpc/internal/transport.newHTTP2Client
	/build/go/src/github.com/prometheus/prometheus/vendor/google.golang.org/grpc/internal/transport/http2_client.go:289 +0xd42

goroutine 6 [syscall, 28589 minutes]:
os/signal.signal_recv(0x0)
	/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-go-1.12.16/share/go/src/runtime/sigqueue.go:139 +0x9c
os/signal.loop()
	/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-go-1.12.16/share/go/src/os/signal/signal_unix.go:23 +0x22
created by os/signal.init.0
	/nix/store/eeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee-go-1.12.16/share/go/src/os/signal/signal_unix.go:29 +0x41
...
```

For more see the attached file, please.

**Comment by hawran** [comment]
**?**
I've just tried do attach the zipped file:
```
$ lsal *.zip
-r--r--r-- 1 ... ... 273226 Mar 17 14:15 200317-135331.prometheus-1.--previous.log.reduced.log.zip
$ 
```
 and got an error as follows (see attached (if it not fails again)):
![Screenshot from 2020-03-17 14-18-08](https://user-images.githubusercontent.com/25924718/76860133-cd5d2380-685a-11ea-9d25-c07fdbe1d948.png)

**Comment by hawran** [comment]
Still cannot upload the file, what's wrong?

**Comment by roidelapluie** [comment]
Cah you paste the last lines in a github gist?

**Comment by hawran** [comment]
> Cah you paste the last lines in a github gist?

Well, how many is "the last lines"?

**Comment by roidelapluie** [comment]
The goroutine dump, if possible

**Comment by hawran** [comment]
https://gist.github.com/hawran/c74a82e2910ca33f632aa42bdf747619

**Comment by hawran** [comment]
> The goroutine dump, if possible

?
Which one?
```
$ egrep goroutine 200317-135331.prometheus-1.--previous.log.reduced.log|wc -l
5822
```

**Comment by roidelapluie** [comment]
mmh grep scrapeCache -C 10 200317-135331.prometheus-1.--previous.log.reduced.log

**Comment by roidelapluie** [comment]
As this panic prometheus I am willing to have it in the 2.17 release. However it might be in a 2.17.1 release.

**Comment by hawran** [comment]
> mmh grep scrapeCache -C 10 200317-135331.prometheus-1.--previous.log.reduced.log

Oops, I can see it now, the logs snippet (within my first comment) got messed up, I've fixed it...

**Comment by roidelapluie** [comment]
That is the map read, I do not see the map write

**Comment by hawran** [comment]
> That is the map read, I do not see the map write

This is the the only place I can find `scrapeCache` within logs.
Could you give me some further hint to look up?

**Comment by brian-brazil** [comment]
The writes are in `trackStaleness` and `iterDone`.

I suspect what is happening is that a new loop is created reusing an old scrape cache, but `endOfRunStaleness` has yet to run and is still using the cache.

**Comment by hawran** [comment]
> The writes are in `trackStaleness` and `iterDone`.
> 
> I suspect what is happening is that a new loop is created reusing an old scrape cache, but `endOfRunStaleness` has yet to run and is still using the cache.

OK, thank you.
However, I'd like to provide the whole log file, any idea why I can't attach it here?

**Comment by hawran** [comment]
> > The writes are in `trackStaleness` and `iterDone`.
> > I suspect what is happening is that a new loop is created reusing an old scrape cache, but `endOfRunStaleness` has yet to run and is still using the cache.
> 
> OK, thank you.
> However, I'd like to provide the whole log file, any idea why I can't attach it here?

Meanwhile:
```
$ egrep -i 'trackStaleness|iterDone|endOfRunStaleness' 200317-135331.prometheus-1.--previous.log.reduced.log
github.com/prometheus/prometheus/scrape.(*scrapeLoop).endOfRunStaleness(0xc0f1760240, 0xbf9445895f2e5033, 0x618189ce0ba36, 0x45bf9e0, 0xc16d72bf90, 0x37e11d600)
github.com/prometheus/prometheus/scrape.(*scrapeLoop).endOfRunStaleness(0xc0969c2fc0, 0xbf94458c4345685c, 0x6181b4c394a70, 0x45bf9e0, 0xc3b67a5900, 0x37e11d600)
github.com/prometheus/prometheus/scrape.(*scrapeLoop).endOfRunStaleness(0xc0f1760510, 0xbf94458984123c80, 0x61818bd5f706e, 0x45bf9e0, 0xc20c021a40, 0x37e11d600)
$
```

**Comment by roidelapluie** [comment]
> The writes are in `trackStaleness` and `iterDone`.
> 
> I suspect what is happening is that a new loop is created reusing an old scrape cache, but `endOfRunStaleness` has yet to run and is still using the cache.

I came to this conclusion without reading this comment.

I have then implemented #7011.

I think that it is fine not to track staleness if we pass the scrape to a new loop as the new loop will take care of marking the old metrics as stale.

**Comment by roidelapluie** [comment]
Fixed in 2.17
